app/controllers/api/databases.php

```diff
@@ -2957,45 +3337,100 @@ function () use ($dbForProject, $privateCollectionId, $document, $data) {
             }
         }
 
-        $documentSecurity = $collection->getAttribute('documentSecurity', false);
-        $validator = new Authorization(Database::PERMISSION_DELETE);
-        $valid = $validator->isValid($collection->getDelete());
-        if (!$documentSecurity && !$valid) {
-            throw new Exception(Exception::USER_UNAUTHORIZED);
-        }
-
         // Read permission should not be required for delete
         $document = Authorization::skip(fn() => $dbForProject->getDocument('database_' . $database->getInternalId() . '_collection_' . $collection->getInternalId(), $documentId));
 
         if ($document->isEmpty()) {
             throw new Exception(Exception::DOCUMENT_NOT_FOUND);
         }
 
-        $privateCollectionId = 'database_' . $database->getInternalId() . '_collection_' . $collection->getInternalId();
+        $checkPermissions = function (Document $collection, Document $document) use (&$checkPermissions, $dbForProject, $database) {
+            $documentSecurity = $collection->getAttribute('documentSecurity', false);
+            $validator = new Authorization(Database::PERMISSION_DELETE);
 
-        if ($documentSecurity && !$valid) {
-            try {
-                $dbForProject->withRequestTimestamp($requestTimestamp, function () use ($dbForProject, $privateCollectionId, $documentId) {
-                    return $dbForProject->deleteDocument($privateCollectionId, $documentId);
-                });
-            } catch (AuthorizationException) {
+            $valid = $validator->isValid($collection->getDelete());
+            if (!$documentSecurity && !$valid) {
                 throw new Exception(Exception::USER_UNAUTHORIZED);
             }
-        } else {
-            Authorization::skip(function () use ($dbForProject, $requestTimestamp, $privateCollectionId, $documentId) {
-                return $dbForProject->withRequestTimestamp($requestTimestamp, function () use ($dbForProject, $privateCollectionId, $documentId) {
-                    return $dbForProject->deleteDocument($privateCollectionId, $documentId);
-                });
-            });
-        }
 
-        $dbForProject->deleteCachedDocument($privateCollectionId, $documentId);
+            $valid = $valid || $validator->isValid($document->getDelete());
+            if ($documentSecurity && !$valid) {
+                throw new Exception(Exception::USER_UNAUTHORIZED);
+            }
+
+            $relationships = \array_filter(
+                $collection->getAttribute('attributes', []),
+                fn ($attribute) => $attribute->getAttribute('type') === Database::VAR_RELATIONSHIP
+            );
+
+            foreach ($relationships as $relationship) {
+                $related = $document->getAttribute($relationship->getAttribute('key'));
```

Comment:
I'm getting the following error:

```
appwrite  | [Error] Timestamp: 2023-03-28T20:56:50+00:00
appwrite  | [Error] Method: DELETE
appwrite  | [Error] URL: /v1/databases/:databaseId/collections/:collectionId/documents/:documentId
appwrite  | [Error] Type: Error
appwrite  | [Error] Message: Call to a member function getAttribute() on string
appwrite  | [Error] File: /usr/src/code/app/controllers/api/databases.php
appwrite  | [Error] Line: 3367
```

The document I'm deleting as a single related`movie` and another related `user`.

Reply:
Oh...it looks like it's happening because the $document down below:

```php
foreach ($related as $document) {
```

changes the `$document` from outside the scope.